File: solutions/fable-1/fable/python/calc.py
import sys


def tokenize(s):
    tokens = []
    i = 0
    n = len(s)
    while i < n:
        c = s[i]
        if c.isspace():
            i += 1
        elif c.isdigit():
            j = i
            while j < n and s[j].isdigit():
                j += 1
            tokens.append(int(s[i:j]))
            i = j
        else:
            tokens.append(c)
            i += 1
    return tokens


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def peek(self):
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def next(self):
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def expr(self):
        val = self.term()
        while self.peek() in ('+', '-'):
            op = self.next()
            rhs = self.term()
            if op == '+':
                val += rhs
            else:
                val -= rhs
        return val

    def term(self):
        val = self.factor()
        while self.peek() in ('*', '/'):
            op = self.next()
            rhs = self.factor()
            if op == '*':
                val *= rhs
            else:
                q = abs(val) // abs(rhs)
                if (val < 0) != (rhs < 0):
                    q = -q
                val = q
        return val

    def factor(self):
        tok = self.next()
        if tok == '(':
            val = self.expr()
            if self.peek() == ')':
                self.next()
            return val
        return tok


def evaluate(line):
    tokens = tokenize(line)
    if not tokens:
        return None
    return Parser(tokens).expr()


def main():
    out = []
    for line in sys.stdin.read().split('\n'):
        if not line.strip():
            continue
        out.append(str(evaluate(line)))
    sys.stdout.write('\n'.join(out) + ('\n' if out else ''))


if __name__ == '__main__':
    main()

File: solutions/fable-1/fable/ruby/calc.rb
class Parser
  def initialize(str)
    @tokens = str.scan(/\d+|[-+*\/()]/)
    @pos = 0
  end

  def parse
    v = expr
    v
  end

  private

  def peek
    @tokens[@pos]
  end

  def advance
    t = @tokens[@pos]
    @pos += 1
    t
  end

  def expr
    v = term
    while peek == '+' || peek == '-'
      op = advance
      r = term
      v = op == '+' ? v + r : v - r
    end
    v
  end

  def term
    v = factor
    while peek == '*' || peek == '/'
      op = advance
      r = factor
      if op == '*'
        v *= r
      else
        # truncate toward zero
        q = v.abs / r.abs
        v = (v < 0) ^ (r < 0) ? -q : q
      end
    end
    v
  end

  def factor
    t = advance
    if t == '('
      v = expr
      advance # ')'
      v
    else
      t.to_i
    end
  end
end

out = []
$stdin.each_line do |line|
  s = line.strip
  next if s.empty?
  out << Parser.new(s).parse
end
puts out.join("\n") unless out.empty?
